src/main/java/org/opensearch/neuralsearch/sparse/NeuralSparseIndexShard.java

```diff
@@ -0,0 +1,230 @@
+import org.opensearch.index.engine.Engine;
+import org.opensearch.index.engine.EngineException;
+import org.opensearch.index.shard.IllegalIndexShardStateException;
+import org.opensearch.index.shard.IndexShard;
+import org.opensearch.neuralsearch.sparse.cache.ClusteredPostingCache;
+import org.opensearch.neuralsearch.sparse.cache.ForwardIndexCache;
+import org.opensearch.neuralsearch.sparse.cache.ForwardIndexCacheItem;
+import org.opensearch.neuralsearch.sparse.common.PredicateUtils;
+import org.opensearch.neuralsearch.sparse.cache.CacheKey;
+import org.opensearch.neuralsearch.sparse.cache.CacheGatedForwardIndexReader;
+import org.opensearch.neuralsearch.sparse.cache.CacheGatedPostingsReader;
+import org.opensearch.neuralsearch.sparse.codec.SparseTermsLuceneReader;
+import org.opensearch.neuralsearch.sparse.codec.SparseBinaryDocValuesPassThrough;
+import org.apache.lucene.index.SegmentReadState;
+import org.opensearch.neuralsearch.sparse.mapper.SparseTokensField;
+
+import java.util.Set;
+
+import java.io.IOException;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+/**
+ * NeuralSparseIndexShard wraps IndexShard and adds methods to perform neural-sparse related operations against the shard
+ */
+@Log4j2
+@RequiredArgsConstructor
+public class NeuralSparseIndexShard {
+    @Getter
+    @NonNull
+    private final IndexShard indexShard;
+
+    private static final String WARM_UP_SEARCHER_SOURCE = "warm-up-searcher-source";
+    private static final String CLEAR_CACHE_SEARCHER_SOURCE = "clear-cache-searcher-source";
+
+    /**
+     * Return the name of the shards index
+     *
+     * @return Name of shard's index
+     */
+    public String getIndexName() {
+        return indexShard.shardId().getIndexName();
+    }
+
+    /**
+     * Load all the neural-sparse segments for this shard into the cache.
+     * Preloads sparse field data to improve query performance.
+     * Uses read lock to allow concurrent warmup operations but prevent conflicts with clear cache.
+     * Early stop to save resources if this is a repeated request
+     */
+    public void warmUp() throws IOException {
+        try (Engine.Searcher searcher = indexShard.acquireSearcher(WARM_UP_SEARCHER_SOURCE)) {
+            for (final LeafReaderContext leafReaderContext : searcher.getIndexReader().leaves()) {
+                final LeafReader leafReader = leafReaderContext.reader();
+
+                // Find all sparse FieldInfos in this segment
+                final Set<FieldInfo> sparseFieldInfos = collectSparseFieldInfos(leafReader);
+
+                // Use CacheGated readers to automatically populate cache on read
+                for (FieldInfo fieldInfo : sparseFieldInfos) {
+                    try {
+                        final SegmentReader segmentReader = Lucene.segmentReader(leafReader);
+                        final SegmentInfo segmentInfo = segmentReader.getSegmentInfo().info;
```

Comment:
- sorry I just found out this issue. It should be slightly performant to move segmentReader and segmentInfo outside of FieldInfo for-loop. 
- Also seem not necessary to try catch this for loop, just put caches after `IllegalIndexShardStateException`